File: leet code  Sheet2/java/3. Longest Substring Without Repeating Characters.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Main {
    public static void main(String[] argv)
    {
        Scanner scanner = new Scanner(System.in);

    }

    /*
       s
         e
      01234567
     "abcabcbb"
    */
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        // a = 0
        // b = 1
        // c = 2
        int start = 0;
        int maxSize = 0; // 3
        for(int end =0; end<s.length();end++){ // end = 3
            char c = s.charAt(end); // a
            if (map.containsKey(c) && map.get(c) >= start){
                start = map.get(c) + 1;
            }
            maxSize = Math.max(maxSize, end - start + 1);

            map.put(c, end);
        }

        return maxSize;
    }
}

File: leet code  Sheet2/c++/3. Longest Substring Without Repeating Characters.cpp
#include<iostream>
#include<map>
using namespace std;
//     e
//   s  
//  01234567
// "abcabcbb"
int lengthOfLongestSubstring(string s) {
	
	map<char, int> map;
	// a = 0
	// b = 1
	// c = 2
	int start = 0;
	int maxSize = 0; // 3
	for(int end = 0;end<s.size();end++){ // 0 1 2 3
		char c = s[end]; // a
		
		if (map.find(c) != map.end() && map[c] >= start){
			start = map[c] + 1;
		}
		maxSize = max(maxSize, end - start + 1);
		map[c] = end;
	}
	
	return maxSize;
}
int main(){
	
}
